controllers/remote/keyedmutex.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package remote
+
+import "sync"
+
+// keyedMutex is a mutex locking on the key provided to the Lock function.
+// Only one caller can hold the lock for a specific key at a time.
+// A second Lock call if the lock is already held for a key returns false.
+type keyedMutex struct {
+	locksMtx sync.Mutex
+	locks    map[interface{}]*sync.Mutex
```

Comment:
If we only need to support `TryLock`, we can simplify this further. It should be okay to just track the set of locked keys. Locking on the mutex within the key seems unneccesary, as we fail early already when the key is present.

Reply:
I can see what @fgutmann is saying as well, what it looks like we're really looking for is a locked set of strings. I would go even further and say the set should be very-well typed as a `client.ObjectKey`, and avoid using `interface{}`

Reply:
Good point will do

Reply:
I refactored in a different way. See: https://github.com/kubernetes-sigs/cluster-api/pull/7537#discussion_r1021511554